### PR TITLE
3.0 interests

### PR DESCRIPTION
**Pull request by mariorudloff**
Hi,

i needed the mailchimp interests as checkboxes for my form.
So this includes the following:
- Extendend MailChimp API to load interests based on listeId and categoryId
- Updated compser dependencies so that its clear this package is for the 2.3 LTS Neos version (because 4.0 isnt compatible)
- Included the "additionalFields" Bugfix
- Added a custom form element for the interests handling the types checkbox, radio and dropdown
- Changed all private MailChimpService members to protected to provide a better extensibility

Regards
Mario

### CODE REVIEW

File: Classes/Wwwision/Neos/MailChimp/Domain/Service/MailChimpService.php [hunk 1]
```diff
@@ -24,17 +25,23 @@ class MailChimpService
     /**
      * @var RequestEngineInterface
      */
-    private $requestEngine;
+    protected $requestEngine;
```
Comment (bwaidelich) [inline Classes/Wwwision/Neos/MailChimp/Domain/Service/MailChimpService.php]: Why did you change the visibility modifier here and below? any specific reason?

File: Classes/Wwwision/Neos/MailChimp/Domain/Service/MailChimpService.php [hunk 2]
```diff
@@ -76,6 +83,44 @@ public function getListById($listId)
 
     /**
      * @param string $listId
+     * @param string $interestCategoryId
+     * @return array
+     */
+    public function getCategoryByListIdAndInterestCategoryId($listId, $interestCategoryId)
+    {
+        $cacheKey = "MailChimp_List_" . $listId ."_Category_" . $interestCategoryId;
```
Comment (bwaidelich) [inline Classes/Wwwision/Neos/MailChimp/Domain/Service/MailChimpService.php]: if you use double quotes you can just write it as
```php
"MailChimp_List_${listId}_Category_${interestCategoryId}"
```

File: Classes/Wwwision/Neos/MailChimp/Domain/Service/MailChimpService.php [hunk 3]
```diff
@@ -24,17 +25,23 @@ class MailChimpService
...
      * @var string
      */
-    private $apiEndpoint;
+    protected $apiEndpoint;
+
+    /**
+     * @Flow\Inject(lazy=false)
+     * @var VariableFrontend
+     */
+    protected $cache;
```
Comment (bwaidelich) [inline Classes/Wwwision/Neos/MailChimp/Domain/Service/MailChimpService.php]: I tried to avoid the "magic property injection" in this class.
If we use a cache we should constructor-inject it like the other dependencies

File: Classes/Wwwision/Neos/MailChimp/Domain/Service/MailChimpService.php (hunk 2 above)
Comment (bwaidelich) [inline Classes/Wwwision/Neos/MailChimp/Domain/Service/MailChimpService.php]: Besides we should probably add the vendor namespace (Wwwision) to the cache key to prevent any clashes. And lastly we should hash `listId` and `interestCategoryId` to prevent invalid cache identifiers.

File: Classes/Wwwision/Neos/MailChimp/Domain/Service/MailChimpService.php [hunk 4]
```diff
@@ -76,6 +83,44 @@ public function getListById($listId)
...
+    public function getCategoryByListIdAndInterestCategoryId($listId, $interestCategoryId)
+    {
+        $cacheKey = "MailChimp_List_" . $listId ."_Category_" . $interestCategoryId;
+        if ($this->cache->has($cacheKey)) {
+            return $this->cache->get($cacheKey);
+        }
+
+        $categoryResult = $this->get("lists/$listId/interest-categories/$interestCategoryId");
+
+        $this->cache->set($cacheKey, $categoryResult, [], 60 * 60 * 1); // 1 hour caching
```
Comment (bwaidelich) [inline Classes/Wwwision/Neos/MailChimp/Domain/Service/MailChimpService.php]: 1 hour. Seems a bit arbitrary. Any specific reason for why you chose this duration?

File: Classes/Wwwision/Neos/MailChimp/Domain/Service/MailChimpService.php [hunk 5]
```diff
@@ -76,6 +83,44 @@ public function getListById($listId)
 
     /**
      * @param string $listId
+     * @param string $interestCategoryId
+     * @return array
```
Comment (bwaidelich) [inline Classes/Wwwision/Neos/MailChimp/Domain/Service/MailChimpService.php]: What kind of array will this return? a little more inline docs would be helpful

File: Classes/Wwwision/Neos/MailChimp/Domain/Service/MailChimpService.php [hunk 6]
```diff
@@ -76,6 +83,44 @@ public function getListById($listId)
...
+
+        $this->cache->set($cacheKey, $categoryResult, [], 60 * 60 * 1); // 1 hour caching
+
+        return $categoryResult;
+    }
+
+    /**
+     * @param string $listId
+     * @param string $interestCategoryId
+     * @return array
```
Comment (bwaidelich) [inline Classes/Wwwision/Neos/MailChimp/Domain/Service/MailChimpService.php]: some more context would be nice

File: Classes/Wwwision/Neos/MailChimp/Domain/Service/MailChimpService.php [hunk 7]
```diff
@@ -209,4 +271,22 @@ private function makeRequest($method, $resource, array $arguments = null)
         throw new MailChimpException($errorMessage, 1483533997);
     }
 
+    public function getInterestsFormOptionsByListIdAndInterestCategoryId($listId, $categoryId)
```
Comment (bwaidelich) [inline Classes/Wwwision/Neos/MailChimp/Domain/Service/MailChimpService.php]: Missing doc comments

File: composer.json [hunk 8]
```diff
@@ -4,6 +4,7 @@
     "description": "TYPO3 Flow package that integrates MailChimp into your Neos site / Flow Application",
     "license": "GPL-3.0+",
     "require": {
+        "typo3/neos": "^2.3",
```
Comment (bwaidelich) [inline composer.json]: This package doesn't depend on Neos for a reason ;)
It can be used with Flow alone and I'd like to keep it that way. Any reason why you added this dependency?

Comment (cbunhok) [reply composer.json]: @bwaidelich: I think he might want to fix an issue that cannot load "TYPO3\Neos\Controller\Module\AbstractModuleController" namespace. I've installed this package version 3.0 in my flow application, and I got the above issue. Do you have any ideas?

Comment (bwaidelich) [reply composer.json]: @cbunhok you're right, I wasn't aware that the compiler requires this class - and the same for the Form Finisher that requires `neos/form` to be installed. I see if I can work around this

Comment (bwaidelich) [reply composer.json]: ..There seems to be no easy fix, so 879b815 .. :-/

**Comment by bwaidelich** [comment]
Closing due to missing feedback - sorry, folks. Feel free to re-open or create a new PR with this feature
